### PR TITLE
Organize some of the duplicated TLS code into a separate crate

### PR DESCRIPTION
**Pull request by alexpyattaev**
#### Problem
A whole bunch of places in the code have been defining the same structure SkipServerVerification for use in quic related code. There were at least 3 independent definitions for this struct with zero functional differences in different project locations.

This created unnecessary duplication as well as strange code dependencies where turbine would depend on quic-client for this one struct, even though logically it makes no sense for turbine to depend on quic-client.

#### Summary of Changes

The following moved to tls-utils crate:

SkipServerVerification
SkipClientVerification
new_dummy_x509_certificate

All references to them redirected. Multiple redundant definitions removed.

### CODE REVIEW

File: tls-utils/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+version.workspace = true
+authors.workspace = true
+repository.workspace = true
+homepage.workspace = true
+license.workspace = true
+edition.workspace = true
+
+[dependencies]
+solana-sdk = { workspace = true }
+
```
Comment (KirillLykov) [inline tls-utils/Cargo.toml]: dependencies should be sorted

File: tls-utils/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+rustls = { workspace = true, features=["ring"] }
+x509-parser = { workspace = true }
+solana-keypair = { workspace = true }
+solana-signer = { workspace = true }
+solana-pubkey = { workspace = true }
+
+
```
Comment (KirillLykov) [inline tls-utils/Cargo.toml:20]: extra space

File: tls-utils/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+//! Collection of TLS related code fragments that end up popping up everywhere where quic is used.
+//! Aggregated here to avoid bugs due to conflicting implementations of the same functionality
```
Comment (KirillLykov) [inline tls-utils/src/lib.rs]:
```suggestion
//! Aggregated here to avoid bugs due to conflicting implementations of the same functionality.
```

File: tls-utils/README [hunk 4]
```diff
@@ -0,0 +1 @@
+A collection of utility functions and structures needed to bridge the conceptual gap between conventional TLS security model in protocols like QUIC and what Solana does
```
Comment (KirillLykov) [inline tls-utils/README]:
```suggestion
A collection of utility functions and structures needed to bridge the conceptual gap between conventional TLS security model in protocols like QUIC and what Solana does.
```

**Comment by KirillLykov** [comment]
Looks like a good change to me

File: tls-utils/Cargo.toml (hunk 1 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: done

File: tls-utils/Cargo.toml (hunk 2 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml:20]: done

File: tls-utils/src/lib.rs (hunk 3 above)
Comment (alexpyattaev) [reply tls-utils/src/lib.rs]: corrected

File: tls-utils/Cargo.toml (hunk 1 above)
Comment (KirillLykov) [reply tls-utils/Cargo.toml]: have you pushed your changes? I still see space and unsorted dependencies.

Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: yes, just did

**Comment by KirillLykov** [comment]
I think you can press now "ready for review"

**Comment by alexpyattaev** [comment]
Thanks, will do.

> 
> I think you can press now "ready for review"
>

File: quic-client/Cargo.toml [hunk 5]
```diff
@@ -29,10 +29,10 @@ solana-quic-definitions = { workspace = true }
 solana-rpc-client-api = { workspace = true }
 solana-signer = { workspace = true }
 solana-streamer = { workspace = true }
+solana-tls-utils = { workspace = true }
 solana-transaction-error = { workspace = true }
 thiserror = { workspace = true }
 tokio = { workspace = true, features = ["full"] }
-
```
Comment (gregcusack) [inline quic-client/Cargo.toml]: should be a new line here

File: tls-utils/src/tls_certificates.rs [hunk 6]
```diff
@@ -1,3 +1,4 @@
+#![allow(clippy::arithmetic_side_effects)]
```
Comment (gregcusack) [inline tls-utils/src/tls_certificates.rs]: is this addition needed? isn't the file just moved to a new location?

File: tls-utils/README [hunk 7]
```diff
@@ -0,0 +1 @@
+A collection of utility functions and structures needed to bridge the conceptual gap between conventional TLS security model in protocols like QUIC and what Solana does.
```
Comment (gregcusack) [inline tls-utils/README]:
```suggestion
A collection of utility functions and structures needed to bridge the conceptual gap between conventional TLS security models in protocols like QUIC and Agave use cases
```

File: tls-utils/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "solana-tls-utils"
+description = "Solana TLS utilities"
+documentation = "https://docs.rs/solana-tls-utils"
+
+version.workspace = true
```
Comment (gregcusack) [inline tls-utils/Cargo.toml]:
```suggestion
description = "Solana TLS utilities"
documentation = "https://docs.rs/solana-tls-utils"
publish = false

version.workspace = true
```

**Review by gregcusack — COMMENTED** [review]
just a few initial thoughts on this! thank you for making these changes. making sure the crate isn't published will help get this PR past CI

File: quic-client/tests/quic_client.rs [hunk 9]
```diff
@@ -7,14 +7,14 @@ mod tests {
...
-            QuicClientCertificate, QuicLazyInitializedEndpoint,
+             QuicLazyInitializedEndpoint,
         },
         solana_sdk::{packet::PACKET_DATA_SIZE, signature::Keypair},
         solana_streamer::{
             quic::{QuicServerParams, SpawnServerResult},
             streamer::StakedNodes,
-            tls_certificates::new_dummy_x509_certificate,
         },
+        solana_tls_utils::{QuicClientCertificate,new_dummy_x509_certificate},
```
Comment (lijunwangs) [inline quic-client/tests/quic_client.rs]: space after 'QuicClientCertificate', can you do a "cargo fmt"

File: tls-utils/src/tls_certificates.rs (hunk 6 above)
Comment (alexpyattaev) [reply tls-utils/src/tls_certificates.rs]: Yes, this is necessary. A similar "allow" was on the crate level in the original crate. Clippy gets confused with benign addition for a buffer size computation otherwise. If it is not benign then please let me know why. I've changed the code to appease clippy without the #allow, though I feel like in this case it was overly protective. 

The bigger problem is the streamer crate from which the code was taken, where the same clippy warning is triggered many many times if the allow line is removed. I can look into that in a different PR.

File: tls-utils/Cargo.toml (hunk 8 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: set to publish=false as suggested

File: quic-client/tests/quic_client.rs (hunk 9 above)
Comment (alexpyattaev) [reply quic-client/tests/quic_client.rs]: thanks, cargo fmt done as suggested on all files.

File: quic-client/Cargo.toml (hunk 5 above)
Comment (alexpyattaev) [reply quic-client/Cargo.toml]: addressed.

File: tls-utils/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "solana-tls-utils"
+description = "Solana TLS utilities"
+documentation = "https://docs.rs/solana-tls-utils"
+publish = false
+
```
Comment (KirillLykov) [inline tls-utils/Cargo.toml]: extra line here

File: tls-utils/src/tls_certificates.rs (hunk 6 above)
Comment (gregcusack) [reply tls-utils/src/tls_certificates.rs]: looks like you removed this?

**Comment by alexpyattaev** [comment]
Yes, I have changed the code to remove the #allow directive, even though it was present in the streamer crate, as it could be masking other possible bugs (and still probably does in streamer).

3 Dec 2024 19.19.36 Greg Cusack ***@***.***>:

File: tls-utils/Cargo.toml (hunk 10 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: publish=false will have to go away after this is merged as far as I understand.

**Review by KirillLykov — DISMISSED** [review]

File: tls-utils/Cargo.toml (hunk 10 above)
Comment (gregcusack) [reply tls-utils/Cargo.toml]: ya are we going to publish this on crates.io? or no?

File: tls-utils/src/tls_certificates.rs [hunk 11]
```diff
@@ -29,9 +29,22 @@ pub fn new_dummy_x509_certificate(
...
-    key_pkcs8_der.extend_from_slice(&PKCS8_PREFIX);
-    key_pkcs8_der.extend_from_slice(keypair.secret().as_bytes());
+
+    let key_pkcs8_der = {
+        let keypair_secret_bytes = keypair.secret().as_bytes();
+        let keypair_secret_len = keypair_secret_bytes.len();
+        if keypair_secret_len != 32 {
+            panic!("Unexpected secret key length!");
+        }
+        let need_buffer = PKCS8_PREFIX
```
Comment (lijunwangs) [inline tls-utils/src/tls_certificates.rs]:
```suggestion
        let buffer_size = PKCS8_PREFIX
```

File: tls-utils/src/tls_certificates.rs [hunk 12]
```diff
@@ -29,9 +29,22 @@ pub fn new_dummy_x509_certificate(
...
+            .len()
+            .checked_add(keypair_secret_len) //clippy being overly guarded here but optimizer will elide checked_add
+            .expect("Unexpected secret key length!");
+        let mut key_pkcs8_der = Vec::<u8>::with_capacity(need_buffer);
```
Comment (lijunwangs) [inline tls-utils/src/tls_certificates.rs]:
```suggestion
        let mut key_pkcs8_der = Vec::<u8>::with_capacity(buffer_size);
```

File: tls-utils/Cargo.toml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+version = { workspace = true }
+authors = { workspace = true }
+repository = { workspace = true }
+homepage = { workspace = true }
+license = { workspace = true }
+edition = { workspace = true }
+
+[dependencies]
+
```
Comment (gregcusack) [inline tls-utils/Cargo.toml]: nit: shouldn't be a blank line at line 15

**Review by gregcusack — COMMENTED** [review]
looks good. just need to remove the extra blank line in `tls-utils/Cargo.toml`

File: tls-utils/Cargo.toml (hunk 13 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: corrected

File: tls-utils/src/tls_certificates.rs (hunk 12 above)
Comment (alexpyattaev) [reply tls-utils/src/tls_certificates.rs]: corrected as suggested.

File: tls-utils/src/tls_certificates.rs (hunk 11 above)
Comment (alexpyattaev) [reply tls-utils/src/tls_certificates.rs]: corrected as  suggested.

File: tls-utils/Cargo.toml (hunk 10 above)
Comment (alexpyattaev) [reply tls-utils/Cargo.toml]: I have no idea if anyone would ever use it there. Probably there is no benefit to that in this case.

**Review by lijunwangs — APPROVED** [review]

**Review by gregcusack — APPROVED** [review]
lgtm!

**Comment by joncinque** [comment]
This crate needs to be published -- solana-quic-client depends on it, which is used by the CLI and many other clients. We can also remove the dependency on solana-sdk. I'll put in a PR for all that
